frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+			ensure!(
+				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != stash),
+				Error::<T>::AlreadyHead
+			);
+			Queue::<T>::remove(stash);
+			Ok(())
+		}
+
+		/// Control the operation of this pallet.
+		///
+		/// Dispatch origin must be signed by the [`Config::ControlOrigin`].
+		#[pallet::weight(<T as Config>::WeightInfo::control())]
+		pub fn control(origin: OriginFor<T>, eras_to_check: EraIndex) -> DispatchResult {
+			let _ = T::ControlOrigin::ensure_origin(origin)?;
+			ErasToCheckPerBlock::<T>::put(eras_to_check);
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// process up to `remaining_weight`.
+		///
+		/// Returns the actual weight consumed.
+		///
+		/// Written for readability in mind, not efficiency. For example:
+		///
+		/// 1. We assume this is only ever called once per `on_idle`. This is because we know that
+		/// in all use cases, even a single nominator cannot be unbonded in a single call. Multiple
+		/// calls to this function are thus not needed. 2. We will only mark a staker
+		#[cfg_attr(feature = "runtime-benchmarks", allow(unused_variables))]
+		pub(crate) fn process_head(remaining_weight: Weight) -> Weight {
+			let eras_to_check_per_block = ErasToCheckPerBlock::<T>::get();
+			if eras_to_check_per_block.is_zero() {
+				return T::DbWeight::get().reads(1)
+			}
+
+			// NOTE: here we're assuming that the number of validators has only ever increased,
+			// meaning that the number of exposures to check is either this per era, or less.
+			let validator_count = pallet_staking::ValidatorCount::<T>::get();
+
+			// determine the number of eras to check. This is based on both `ErasToCheckPerBlock`
+			// and `remaining_weight` passed on to us from the runtime executive.
+			let final_eras_to_check = {
+				let worse_weight = |v, u| {
+					<T as Config>::WeightInfo::on_idle_check(v * u)
+						.max(<T as Config>::WeightInfo::on_idle_unstake())
+				};
+				let mut try_eras_to_check = eras_to_check_per_block;
+				while worse_weight(validator_count, try_eras_to_check).any_gt(remaining_weight) {
+					try_eras_to_check.saturating_dec();
+					if try_eras_to_check.is_zero() {
+						log!(debug, "early existing because try_eras_to_check is zero");
+						return T::DbWeight::get().reads(1)
+					}
+				}
+
+				drop(eras_to_check_per_block);
+				try_eras_to_check
+			};
+
+			if <T as pallet_staking::Config>::ElectionProvider::ongoing() {
+				// NOTE: we assume `ongoing` does not consume any weight.
+				// there is an ongoing election -- we better not do anything. Imagine someone is not
+				// exposed anywhere in the last era, and the snapshot for the election is already
+				// taken. In this time period, we don't want to accidentally unstake them.
+				return T::DbWeight::get().reads(2)
+			}
+
+			let UnstakeRequest { stash, mut checked, maybe_pool_id } = match Head::<T>::take()
+				.or_else(|| {
+					Queue::<T>::drain()
```

Comment:
Its not very clear to me why we are draining the queue here. May be some comments would help?

Reply:
`::drain` is the same as `::iter()`, but any element that's visited is also removed. 

https://paritytech.github.io/substrate/master/frame_support/storage/trait.IterableStorageMap.html#tymethod.drain